source/Octopus.Tentacle/Diagnostics/Metrics/PersistenceProvider.cs

```diff
@@ -0,0 +1,44 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using k8s.Models;
+using Octopus.Diagnostics;
+using Octopus.Tentacle.Kubernetes;
+
+namespace Octopus.Tentacle.Diagnostics.Metrics
+{
+    public interface IPersistenceProvider
+    {
+        string GetValue(string key);
+        void PersistValue(string key, string value);
+    }
+
+    public class PersistenceProvider : IPersistenceProvider
+    {
+        readonly string configMapName;
+        readonly IKubernetesConfigMapService configMapService;
+        readonly Lazy<V1ConfigMap> metricsConfigMap;
+        readonly ISystemLog log;
+        IDictionary<string, string> ConfigMapData => metricsConfigMap.Value.Data ??= new Dictionary<string, string>();
+
+        public PersistenceProvider(string configMapName, IKubernetesConfigMapService configMapService, ISystemLog log)
+        {
+            this.configMapService = configMapService;
+            this.log = log;
+            this.configMapName = configMapName;
+            metricsConfigMap = new Lazy<V1ConfigMap>(() => configMapService.TryGet(this.configMapName, CancellationToken.None).GetAwaiter().GetResult()
+                ?? throw new InvalidOperationException($"Unable to retrieve Tentacle Configuration from config map for namespace {KubernetesConfig.Namespace}"));
```

Comment:
Does this exception get propagated out of the constructor?

Reply:
No - it gets propagated when the lazy is evaluated.

Specifically - if the lazy fails on the first execution - the same exception is thrown for every attempt.